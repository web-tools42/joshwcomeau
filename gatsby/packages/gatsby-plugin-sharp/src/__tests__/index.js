const path = require(`path`)
const sharp = require(`sharp`)
const fs = require(`fs-extra`)
jest.mock(`../scheduler`)

jest.mock(`async/queue`, () => () => {
  return {
    push: jest.fn(),
  }
})

const { scheduleJob } = require(`../scheduler`)
scheduleJob.mockReturnValue(Promise.resolve())
fs.ensureDirSync = jest.fn()
fs.existsSync = jest.fn().mockReturnValue(false)
let isolatedQueueImageResizing
jest.isolateModules(() => {
  isolatedQueueImageResizing = require(`../index`).queueImageResizing
})

const {
  base64,
  fluid,
  fixed,
  queueImageResizing,
  getImageSize,
  stats,
} = require(`../`)

jest.mock(`gatsby-cli/lib/reporter`, () => {
  return {
    log: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
    activityTimer: () => {
      return {
        start: jest.fn(),
        setStatus: jest.fn(),
        end: jest.fn(),
      }
    },
  }
})

describe(`gatsby-plugin-sharp`, () => {
  const args = {
    duotone: false,
    grayscale: false,
    rotate: false,
  }
  const absolutePath = path.join(__dirname, `images/test.png`)
  const file = getFileObject(absolutePath)

  // used to find all breakpoints in a srcSet string
  const findAllBreakpoints = srcSet => {
    // RegEx to find all occurrences of 'Xw', where 'X' can be any int
    const regEx = /[0-9]+w/g
    return srcSet.match(regEx)
  }

  describe(`queueImageResizing`, () => {
    beforeEach(() => {
      scheduleJob.mockClear()
      fs.existsSync.mockReset()
      fs.existsSync.mockReturnValue(false)
    })

    it(`should round height when auto-calculated`, () => {
      // Resize 144-density.png (281x136) with a 3px width
      const result = queueImageResizing({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: { width: 3 },
      })

      // Width should be: w = (3 * 136) / 281 = 1.451957295
      // We expect value to be rounded to 1
      expect(result.height).toBe(1)
    })

    it(`file name works with spaces & special characters`, async () => {
      // test name encoding with various characters
      const testName = `spaces and '"@#$%^&,`

      const queueResult = queueImageResizing({
        file: getFileObject(
          path.join(__dirname, `images/144-density.png`),
          testName
        ),
        args: { width: 3 },
      })

      const queueResultName = path.parse(queueResult.src).name

      // decoding to check for outputting same name
      expect(decodeURIComponent(queueResultName)).toBe(testName)

      // regex for special characters above and spaces
      // testname should match, the queue result should not
      expect(testName.match(/[!@#$^&," ]/)).not.toBe(false)
      expect(queueResultName.match(/[!@#$^&," ]/)).not.toBe(true)
    })

    // re-enable when image processing on demand is implemented
    it.skip(`should process immediately when asked`, async () => {
      const result = queueImageResizing({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: { width: 3 },
      })

      await result.finishedPromise

      expect(scheduleJob).toMatchSnapshot()
    })

    it(`Shouldn't schedule a job when outputFile already exists`, async () => {
      fs.existsSync.mockReturnValue(true)

      const result = queueImageResizing({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: { width: 3 },
      })

      await result.finishedPromise

      expect(fs.existsSync).toHaveBeenCalledWith(result.absolutePath)
      expect(scheduleJob).not.toHaveBeenCalled()
    })

    it(`Shouldn't schedule a job when with same outputFile is already being queued`, async () => {
      const result = isolatedQueueImageResizing({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: { width: 5 },
      })
      isolatedQueueImageResizing({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: { width: 5 },
      })

      await result.finishedPromise

      expect(scheduleJob).toHaveBeenCalledTimes(1)
    })
  })

  describe(`fluid`, () => {
    it(`includes responsive image properties, e.g. sizes, srcset, etc.`, async () => {
      const result = await fluid({ file })

      expect(result).toMatchSnapshot()
    })

    it(`adds pathPrefix if defined`, async () => {
      const pathPrefix = `/blog`
      const result = await fluid({
        file,
        args: {
          pathPrefix,
        },
      })

      expect(result.src.indexOf(pathPrefix)).toBe(0)
      expect(result.srcSet.indexOf(pathPrefix)).toBe(0)
    })

    it(`keeps original file name`, async () => {
      const result = await fluid({
        file,
      })

      expect(path.parse(result.src).name).toBe(file.name)
      expect(path.parse(result.srcSet).name).toBe(file.name)
    })

    it(`does not change the arguments object it is given`, async () => {
      const args = { maxWidth: 400 }
      await fluid({
        file,
        args,
      })

      expect(args).toEqual({ maxWidth: 400 })
    })

    it(`infers the maxWidth if only maxHeight is given`, async () => {
      const args = { maxHeight: 20 }
      const result = await fluid({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      expect(result.presentationWidth).toEqual(41)
    })

    it(`should throw if maxWidth is less than 1`, async () => {
      const args = { maxWidth: 0 }
      const result = fluid({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      await expect(result).rejects.toThrow()
    })

    it(`accepts srcSet breakpoints`, async () => {
      const srcSetBreakpoints = [50, 70, 150, 250]
      const args = { srcSetBreakpoints }
      const result = await fluid({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      // width of the image tested
      const originalWidth = 281
      const expected = srcSetBreakpoints.map(size => `${size}w`)
      // add the original size of `144-density.png`
      expected.push(`${originalWidth}w`)

      const actual = findAllBreakpoints(result.srcSet)
      // should contain all requested sizes as well as the original size
      expect(actual).toEqual(expect.arrayContaining(expected))
    })

    it(`should throw on srcSet breakpoints less than 1`, async () => {
      const srcSetBreakpoints = [50, 0]
      const args = { srcSetBreakpoints }
      const result = fluid({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      await expect(result).rejects.toThrow()
    })

    it(`ensure maxWidth is in srcSet breakpoints`, async () => {
      const srcSetBreakpoints = [50, 70, 150]
      const maxWidth = 200
      const args = {
        maxWidth,
        srcSetBreakpoints,
      }
      const result = await fluid({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      expect(result.srcSet).toEqual(expect.stringContaining(`${maxWidth}w`))
    })

    it(`reject any breakpoints larger than the original width`, async () => {
      const srcSetBreakpoints = [
        50,
        70,
        150,
        250,
        300, // this shouldn't be in the output as it's wider than the original
      ]
      const maxWidth = 500 // this also shouldn't be in the output
      const args = {
        maxWidth,
        srcSetBreakpoints,
      }
      const result = await fluid({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      // width of the image tested
      const originalWidth = 281
      const expected = srcSetBreakpoints
        // filter out the widths that are larger than the source image width
        .filter(size => size < originalWidth)
        .map(size => `${size}w`)
      // add the original size of `144-density.png`
      expected.push(`${originalWidth}w`)

      const actual = findAllBreakpoints(result.srcSet)
      // should contain all requested sizes as well as the original size
      expect(actual).toEqual(expect.arrayContaining(expected))
      // should contain no other sizes
      expect(actual.length).toEqual(expected.length)
    })

    it(`prevents duplicate breakpoints`, async () => {
      const srcSetBreakpoints = [50, 50, 100, 100, 100, 250, 250]
      const maxWidth = 100
      const args = {
        maxWidth,
        srcSetBreakpoints,
      }
      const result = await fluid({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      const originalWidth = 281
      const expected = [`50w`, `100w`, `250w`, `${originalWidth}w`]

      const actual = findAllBreakpoints(result.srcSet)
      expect(actual).toEqual(expect.arrayContaining(expected))
      expect(actual.length).toEqual(expected.length)
    })
  })

  describe(`fixed`, () => {
    it(`does not warn when the requested width is equal to the image width`, async () => {
      console.warn = jest.fn()
      const args = { width: 1 }

      const result = await fixed({
        file,
        args,
      })

      expect(result.width).toEqual(1)
      expect(console.warn).toHaveBeenCalledTimes(0)
      console.warn.mockClear()
    })

    it(`warns when the requested width is greater than the image width`, async () => {
      console.warn = jest.fn()
      const { width } = await sharp(file.absolutePath).metadata()
      const args = { width: width * 2 }

      const result = await fixed({
        file,
        args,
      })

      expect(result.width).toEqual(width)
      expect(console.warn).toHaveBeenCalledTimes(1)
      console.warn.mockClear()
    })

    it(`correctly infers the width when only the height is given`, async () => {
      const args = { height: 10 }

      const result = await fixed({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args,
      })

      expect(result.width).toEqual(21)
    })
  })

  describe(`base64`, () => {
    it(`converts image to base64`, async () => {
      const result = await base64({
        file,
        args,
      })

      expect(result).toMatchSnapshot()
    })
  })

  describe(`image quirks`, () => {
    // issue https://github.com/nodeca/probe-image-size/issues/20
    it(`handles padding bytes correctly`, () => {
      const result = getImageSize(
        getFileObject(path.join(__dirname, `images/padding-bytes.jpg`))
      )

      expect(result).toMatchSnapshot()
    })
  })

  describe(`tracedSVG`, () => {
    it(`doesn't always run`, async () => {
      const args = {
        maxWidth: 100,
        width: 100,
        tracedSVG: { color: `#FF0000` },
      }

      let result = await fixed({
        file,
        args,
      })

      expect(result.tracedSVG).toBeUndefined()

      result = await fluid({
        file,
        args,
      })

      expect(result.tracedSVG).toBeUndefined()
    })

    it(`runs on demand`, async () => {
      const args = {
        maxWidth: 100,
        width: 100,
        generateTracedSVG: true,
        tracedSVG: { color: `#FF0000` },
        base64: false,
      }

      const fixedSvg = await fixed({
        file,
        args,
      })

      expect(fixedSvg).toMatchSnapshot()

      const fluidSvg = await fluid({
        file,
        args,
      })

      expect(fluidSvg).toMatchSnapshot()
    })
  })

  describe(`duotone`, () => {
    const args = {
      maxWidth: 100,
      width: 100,
      duotone: { highlight: `#ffffff`, shadow: `#cccccc`, opacity: 50 },
    }

    it(`fixed`, async () => {
      let result = await fixed({ file, args })
      expect(result).toMatchSnapshot()
    })

    it(`fluid`, async () => {
      let result = await fluid({ file, args })
      expect(result).toMatchSnapshot()
    })
  })

  describe(`stats`, () => {
    it(`determines if the image is transparent, based on the presence and use of alpha channel`, async () => {
      const result = await stats({ file, args })
      expect(result).toMatchSnapshot()
      expect(result.isTransparent).toEqual(false)

      const alphaResult = await stats({
        file: getFileObject(path.join(__dirname, `images/alphatest.png`)),
        args,
      })
      expect(alphaResult).toMatchSnapshot()
      expect(alphaResult.isTransparent).toEqual(true)
    })
  })
})

function getFileObject(absolutePath, name = `test`) {
  return {
    id: `${absolutePath} absPath of file`,
    name: name,
    absolutePath,
    extension: `png`,
    internal: {
      contentDigest: `1234`,
    },
  }
}
