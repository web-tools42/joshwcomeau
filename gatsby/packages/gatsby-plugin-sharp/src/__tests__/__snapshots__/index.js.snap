// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-sharp base64 converts image to base64 1`] = `
Object {
  "aspectRatio": 1,
  "height": 20,
  "originalName": undefined,
  "src": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsSAAALEgHS3X78AAABKklEQVQ4y8WUy07DMBBFrwMpr4aybjepUeqIFd/CAgmJsuIHgWXVL+Gx4C3xCdQJnpm0UDJNAQnVuouMM8dzY4+DcoA/C/8Fe9bv4ABMMpQOZc5yFPqfwEUQY499jHoY9/BsKQyTRTNMyzu8WAwTbBnI2DY438XbPr0qFsFCvlrkMTERYFgRL3HYUvh5OMdpQqkbBtPC9NDiINQPCZM6LGVvUzIpwNch4V5EGxHS/DeYtjfHZbcyXB8yGfaPimcafLEMHqmw2L5OsWkUUuY6Ee5V28WUP25TXjwPS3iWUFnfcFQPFna98rlmSGL4IOZucQvgT76Po53qeOTYTtp4qpFKe/qM29PhJsVVl3SXsluntLfS7rPvn10MP1Aau+lKBuA9I/kV/AyW6gNTIWTuA/r8lQAAAABJRU5ErkJggg==",
  "width": 20,
}
`;

exports[`gatsby-plugin-sharp duotone fixed 1`] = `
Object {
  "aspectRatio": 1,
  "base64": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAAH6ji2bAAAACXBIWXMAAAPoAAAD6AG1e1JrAAABk0lEQVQ4y7VUy07DMBDsf9IDhcauaVM4IajUIuALOFRI/R24wB8AgkpInJKI9ME5zDi24rx5HjZKvOP17Hg2neh8nNjo8CGESGL7sbpwMvrlaepriF4J3dQLMlsXvzbVFCKHLJ3J8+5PRxqlF0OzleFLkSG7Xrp4faiyRaI/Lgs13QhNVB5uCUyUTDyUloidvkjezvyMEB8B4uZkqAHCiR44bYtCcecSirD5XQDmIGzVjOrEiJo4fhvY2jWTG3C6g/TLmZ+8VwEJekWy20+loUSzA5kDp15BKJmXhmD6Iqcjtep5eSArk0ZQPPpqPNAaWiBpbKt0pLgEU2xuIOewTh7rmFWFRH9/M78uaG8yNuox4oYbbrQEhaI4HKEh7mvfS2OE98WR0r6P26zj2ofW5rWowlzY2WduAQxZ14odmNZuMWA0h6wo5pplDxj+MrgnqDM3W32E/QYNxaxN2f7ztNx6qWUCOC3H+F/Y35EU2XCw3QkGhBrHX/WXLbwG4wewoAwMst84t/0jH4ZG36DFLv9m7E/1HjvIa/canAAAAABJRU5ErkJggg==",
  "height": 100,
  "originalName": undefined,
  "src": "/static/1234/a2779/test.png",
  "srcSet": "/static/1234/a2779/test.png 1x",
  "tracedSVG": undefined,
  "width": 100,
}
`;

exports[`gatsby-plugin-sharp duotone fluid 1`] = `
Object {
  "aspectRatio": 1,
  "base64": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAAH6ji2bAAAACXBIWXMAAAPoAAAD6AG1e1JrAAABk0lEQVQ4y7VUy07DMBDsf9IDhcauaVM4IajUIuALOFRI/R24wB8AgkpInJKI9ME5zDi24rx5HjZKvOP17Hg2neh8nNjo8CGESGL7sbpwMvrlaepriF4J3dQLMlsXvzbVFCKHLJ3J8+5PRxqlF0OzleFLkSG7Xrp4faiyRaI/Lgs13QhNVB5uCUyUTDyUloidvkjezvyMEB8B4uZkqAHCiR44bYtCcecSirD5XQDmIGzVjOrEiJo4fhvY2jWTG3C6g/TLmZ+8VwEJekWy20+loUSzA5kDp15BKJmXhmD6Iqcjtep5eSArk0ZQPPpqPNAaWiBpbKt0pLgEU2xuIOewTh7rmFWFRH9/M78uaG8yNuox4oYbbrQEhaI4HKEh7mvfS2OE98WR0r6P26zj2ofW5rWowlzY2WduAQxZ14odmNZuMWA0h6wo5pplDxj+MrgnqDM3W32E/QYNxaxN2f7ztNx6qWUCOC3H+F/Y35EU2XCw3QkGhBrHX/WXLbwG4wewoAwMst84t/0jH4ZG36DFLv9m7E/1HjvIa/canAAAAABJRU5ErkJggg==",
  "density": 72,
  "originalImg": "/static/1234/a2779/test.png",
  "originalName": undefined,
  "presentationHeight": 100,
  "presentationWidth": 100,
  "sizes": "(max-width: 100px) 100vw, 100px",
  "src": "/static/1234/a2779/test.png",
  "srcSet": "/static/1234/2d2c2/test.png 25w,
/static/1234/84325/test.png 50w,
/static/1234/a2779/test.png 100w",
  "srcSetType": "image/png",
  "tracedSVG": undefined,
}
`;

exports[`gatsby-plugin-sharp fluid includes responsive image properties, e.g. sizes, srcset, etc. 1`] = `
Object {
  "aspectRatio": 1,
  "base64": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsSAAALEgHS3X78AAABKklEQVQ4y8WUy07DMBBFrwMpr4aybjepUeqIFd/CAgmJsuIHgWXVL+Gx4C3xCdQJnpm0UDJNAQnVuouMM8dzY4+DcoA/C/8Fe9bv4ABMMpQOZc5yFPqfwEUQY499jHoY9/BsKQyTRTNMyzu8WAwTbBnI2DY438XbPr0qFsFCvlrkMTERYFgRL3HYUvh5OMdpQqkbBtPC9NDiINQPCZM6LGVvUzIpwNch4V5EGxHS/DeYtjfHZbcyXB8yGfaPimcafLEMHqmw2L5OsWkUUuY6Ee5V28WUP25TXjwPS3iWUFnfcFQPFna98rlmSGL4IOZucQvgT76Po53qeOTYTtp4qpFKe/qM29PhJsVVl3SXsluntLfS7rPvn10MP1Aau+lKBuA9I/kV/AyW6gNTIWTuA/r8lQAAAABJRU5ErkJggg==",
  "density": 72,
  "originalImg": "/static/1234/c616f/test.png",
  "originalName": undefined,
  "presentationHeight": 100,
  "presentationWidth": 100,
  "sizes": "(max-width: 100px) 100vw, 100px",
  "src": "/static/1234/c616f/test.png",
  "srcSet": "/static/1234/c616f/test.png 100w",
  "srcSetType": "image/png",
  "tracedSVG": undefined,
}
`;

exports[`gatsby-plugin-sharp image quirks handles padding bytes correctly 1`] = `
Object {
  "hUnits": "px",
  "height": 1000,
  "mime": "image/jpeg",
  "type": "jpg",
  "wUnits": "px",
  "width": 746,
}
`;

exports[`gatsby-plugin-sharp stats determines if the image is transparent, based on the presence and use of alpha channel 1`] = `
Object {
  "isTransparent": false,
}
`;

exports[`gatsby-plugin-sharp stats determines if the image is transparent, based on the presence and use of alpha channel 2`] = `
Object {
  "isTransparent": true,
}
`;

exports[`gatsby-plugin-sharp tracedSVG runs on demand 1`] = `
Object {
  "aspectRatio": 1,
  "base64": undefined,
  "height": 100,
  "originalName": undefined,
  "src": "/static/1234/c616f/test.png",
  "srcSet": "/static/1234/c616f/test.png 1x",
  "tracedSVG": "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20width='100'%20height='100'%3e%3cpath%20d='M41%2024c-18%207-24%2029-11%2043%2015%2017%2044%208%2046-15%201-19-17-34-35-28'%20fill='red'%20fill-rule='evenodd'/%3e%3c/svg%3e",
  "width": 100,
}
`;

exports[`gatsby-plugin-sharp tracedSVG runs on demand 2`] = `
Object {
  "aspectRatio": 1,
  "base64": undefined,
  "density": 72,
  "originalImg": "/static/1234/c616f/test.png",
  "originalName": undefined,
  "presentationHeight": 100,
  "presentationWidth": 100,
  "sizes": "(max-width: 100px) 100vw, 100px",
  "src": "/static/1234/c616f/test.png",
  "srcSet": "/static/1234/db86b/test.png 25w,
/static/1234/356f1/test.png 50w,
/static/1234/c616f/test.png 100w",
  "srcSetType": "image/png",
  "tracedSVG": "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20width='100'%20height='100'%3e%3cpath%20d='M41%2024c-18%207-24%2029-11%2043%2015%2017%2044%208%2046-15%201-19-17-34-35-28'%20fill='red'%20fill-rule='evenodd'/%3e%3c/svg%3e",
}
`;
